File: sdk/ts/src/emitters/buffering.ts
/**
 * BufferingEmitter — wraps a downstream {@link Emitter} and buffers
 * receipts in memory, flushing them on a configurable interval or batch
 * size.
 *
 * The contract with the downstream emitter is per-receipt, NOT batched —
 * a flush calls `inner.emit(receipt)` once per buffered receipt.
 *
 * !!! CRASH-LOSS RISK !!!
 * Buffered receipts are lost if the process exits before {@link flush}
 * completes. This emitter is NOT suitable for environments where audit
 * completeness is critical. Use a synchronous {@link HttpEmitter} (or a
 * persistent WAL — tracked separately) when every receipt must reach the
 * collector.
 */

import type { AgentReceipt } from "../receipt/types.js";
import type { Emitter } from "./types.js";

export interface BufferingEmitterConfig {
	inner: Emitter;
	/** Flush when the buffer reaches this size. Must be >= 1. */
	maxBatchSize: number;
	/** Flush every N milliseconds while receipts are buffered. Must be >= 1. */
	flushIntervalMs: number;
}

export class BufferingEmitter implements Emitter {
	private readonly inner: Emitter;
	private readonly maxBatchSize: number;
	private readonly flushIntervalMs: number;

	private readonly buffer: AgentReceipt[] = [];
	private timer: ReturnType<typeof setTimeout> | null = null;
	private closed = false;
	// Serialise concurrent flush() calls so the downstream sees one ordered
	// sequence even when emit() and an interval tick race.
	private flushChain: Promise<void> = Promise.resolve();

	constructor(config: BufferingEmitterConfig) {
		if (config.maxBatchSize < 1) {
			throw new Error("BufferingEmitter: maxBatchSize must be >= 1");
		}
		if (config.flushIntervalMs < 1) {
			throw new Error("BufferingEmitter: flushIntervalMs must be >= 1");
		}
		this.inner = config.inner;
		this.maxBatchSize = config.maxBatchSize;
		this.flushIntervalMs = config.flushIntervalMs;
	}

	async emit(receipt: AgentReceipt): Promise<void> {
		if (this.closed) {
			throw new Error("BufferingEmitter: closed");
		}
		this.buffer.push(receipt);
		if (this.buffer.length >= this.maxBatchSize) {
			await this.flush();
			return;
		}
		this.scheduleTimer();
	}

	/**
	 * Drain the buffer through the downstream emitter, one receipt at a
	 * time. Resolves once every buffered receipt has been delivered (or the
	 * downstream has thrown for one). If multiple calls overlap they are
	 * serialised — the second await observes the first call's completion.
	 */
	flush(): Promise<void> {
		const next = this.flushChain.then(() => this.doFlush());
		this.flushChain = next.then(
			() => {},
			() => {},
		);
		return next;
	}

	/**
	 * Stop the interval timer and flush the remaining buffer. After close
	 * subsequent {@link emit} calls throw.
	 */
	async close(): Promise<void> {
		if (this.closed) {
			return;
		}
		this.closed = true;
		this.cancelTimer();
		await this.flush();
	}

	private async doFlush(): Promise<void> {
		this.cancelTimer();
		// Splice once — anything emitted while we're flushing goes into the
		// next batch, not this one. This keeps the chain progressing even if
		// the downstream is slow.
		const batch = this.buffer.splice(0, this.buffer.length);
		const errors: unknown[] = [];
		for (const receipt of batch) {
			// Attempt every receipt — failing fast on the first error would
			// silently drop receipts already removed from the buffer.
			try {
				await this.inner.emit(receipt);
			} catch (err) {
				errors.push(err);
			}
		}
		if (errors.length === 1) {
			throw errors[0];
		}
		if (errors.length > 1) {
			throw new AggregateError(
				errors,
				`BufferingEmitter: ${errors.length} of ${batch.length} receipts failed`,
			);
		}
	}

	private scheduleTimer(): void {
		if (this.timer !== null) {
			return;
		}
		this.timer = setTimeout(() => {
			this.timer = null;
			// Swallow errors from the timer-driven flush: they would otherwise
			// become unhandled promise rejections. Real callers should rely on
			// the explicit flush()/close() return values for error surfacing.
			void this.flush().catch(() => {});
		}, this.flushIntervalMs);
		// Don't keep the Node event loop alive solely for the flush timer.
		const t = this.timer as { unref?: () => void };
		if (typeof t.unref === "function") {
			t.unref();
		}
	}

	private cancelTimer(): void {
		if (this.timer !== null) {
			clearTimeout(this.timer);
			this.timer = null;
		}
	}
}

File: sdk/go/emitters/buffering.go
package emitters

import (
	"context"
	"errors"
	"sync"
	"time"

	"github.com/agent-receipts/ar/sdk/go/receipt"
)

// BufferingEmitterConfig configures [BufferingEmitter].
type BufferingEmitterConfig struct {
	// Inner is the downstream emitter that receives per-receipt deliveries
	// when the buffer flushes. Required.
	Inner Emitter

	// MaxBatchSize triggers a flush as soon as the buffer reaches this
	// size. Must be >= 1.
	MaxBatchSize int

	// FlushInterval triggers a flush after this duration of buffered
	// inactivity. Must be > 0.
	FlushInterval time.Duration
}

// BufferingEmitter wraps a downstream [Emitter] and buffers receipts in
// memory, flushing them on a configurable interval or batch size.
//
// The contract with the downstream emitter is PER-RECEIPT, not batched:
// a flush calls Inner.Emit once per buffered receipt.
//
// !!! CRASH-LOSS RISK !!!
// Buffered receipts are lost if the process exits before [BufferingEmitter.Flush]
// completes. This emitter is NOT suitable for environments where audit
// completeness is critical. Use a synchronous [HttpEmitter] (or a
// persistent WAL — tracked separately) when every receipt must reach the
// collector.
type BufferingEmitter struct {
	inner         Emitter
	maxBatchSize  int
	flushInterval time.Duration

	mu     sync.Mutex
	buffer []receipt.AgentReceipt
	timer  *time.Timer
	closed bool
}

// NewBuffering constructs a [BufferingEmitter] from the given config.
// Returns an error for invalid sizes/intervals so a misconfiguration is
// caught at construction rather than on the first Emit.
func NewBuffering(cfg BufferingEmitterConfig) (*BufferingEmitter, error) {
	if cfg.Inner == nil {
		return nil, errors.New("BufferingEmitter: Inner is required")
	}
	if cfg.MaxBatchSize < 1 {
		return nil, errors.New("BufferingEmitter: MaxBatchSize must be >= 1")
	}
	if cfg.FlushInterval <= 0 {
		return nil, errors.New("BufferingEmitter: FlushInterval must be > 0")
	}
	return &BufferingEmitter{
		inner:         cfg.Inner,
		maxBatchSize:  cfg.MaxBatchSize,
		flushInterval: cfg.FlushInterval,
	}, nil
}

// Emit appends r to the buffer and either flushes immediately (when
// MaxBatchSize is reached) or schedules a flush after FlushInterval.
func (b *BufferingEmitter) Emit(ctx context.Context, r receipt.AgentReceipt) error {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return errors.New("BufferingEmitter: closed")
	}
	b.buffer = append(b.buffer, r)
	if len(b.buffer) >= b.maxBatchSize {
		batch := b.takeBatchLocked()
		b.mu.Unlock()
		return b.deliver(ctx, batch)
	}
	b.scheduleTimerLocked()
	b.mu.Unlock()
	return nil
}

// Flush delivers every buffered receipt through the inner emitter.
// Resolves once each one has been delivered (or the inner emitter
// returned an error for one of them).
func (b *BufferingEmitter) Flush(ctx context.Context) error {
	b.mu.Lock()
	batch := b.takeBatchLocked()
	b.mu.Unlock()
	return b.deliver(ctx, batch)
}

// Close stops the interval timer and flushes the remaining buffer. After
// Close, subsequent Emit calls return an error. Safe to call multiple
// times.
func (b *BufferingEmitter) Close(ctx context.Context) error {
	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return nil
	}
	b.closed = true
	batch := b.takeBatchLocked()
	b.mu.Unlock()
	return b.deliver(ctx, batch)
}

// ----------------------------------------------------------------------

// takeBatchLocked drains the buffer and cancels any pending timer.
// Caller must hold b.mu.
func (b *BufferingEmitter) takeBatchLocked() []receipt.AgentReceipt {
	if b.timer != nil {
		b.timer.Stop()
		b.timer = nil
	}
	if len(b.buffer) == 0 {
		return nil
	}
	batch := b.buffer
	b.buffer = nil
	return batch
}

func (b *BufferingEmitter) scheduleTimerLocked() {
	if b.timer != nil {
		return
	}
	b.timer = time.AfterFunc(b.flushInterval, b.onTimer)
}

func (b *BufferingEmitter) onTimer() {
	b.mu.Lock()
	b.timer = nil
	if b.closed {
		b.mu.Unlock()
		return
	}
	batch := b.buffer
	b.buffer = nil
	b.mu.Unlock()
	if len(batch) == 0 {
		return
	}
	// Swallow errors on the timer path so an internal goroutine never
	// crashes the host program on a downstream failure. Callers should
	// rely on Flush / Close for error surfacing.
	_ = b.deliver(context.Background(), batch)
}

func (b *BufferingEmitter) deliver(ctx context.Context, batch []receipt.AgentReceipt) error {
	for _, r := range batch {
		if err := b.inner.Emit(ctx, r); err != nil {
			return err
		}
	}
	return nil
}
